Prometheus endpoint for metrics

klaren opened the pull request:
[Prometheus](https://prometheus.io/) is monitoring and alerting toolkit maintained by the [Cloud Native Computing Foundation](https://www.cncf.io/).

To enable the Prometheus HTTP endpoint you have to set `metrics.prometheus.enabled=true` in your `neo4j.conf`. The setting `metrics.prometheus.endpoint=<ip:port>` can be used to change the listening ip and port, default one is `localhost:2004`.

For example, adding the following to your configuration file:
```
metrics.prometheus.enabled=true
metrics.prometheus.endpoint=localhost:2004
```
will start a Prometheus endpoint at port 2004 accepting connections from the local machine.

Security wise, Prometheus states the following:
>It is presumed that untrusted users have access to the Prometheus HTTP endpoint and logs. They have access to all time series information contained in the database, plus a variety of operational/debugging information.

and should thus be managed at a firewall/OS level. If authentication is a requirement, a reverse proxy can be used on the local machine to expose the information.

klaren committed:
Add a prometheus endpoint to metrics module

klaren committed:
Added integration test for Prometheus endpoint

klaren committed:
Updated license files

klaren committed:
Extracted Prometheus dependency version to variable

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/MetricsSettings.java:
```diff
@@ -153,4 +153,12 @@
+
+    @Description( "The hostname and port to use as Prometheus endpoint" )
+    public static Setting<HostnamePort> prometheusEndpoint =
+            setting( "metrics.prometheus.endpoint", HOSTNAME_PORT, "localhost:8080" );
```
wouldn't say that 8080 is a good default port value, why not to use next port after default graphite server (2004)?

klaren replied:
It's an HTTP server, they typically reside in the 8000 range. 8080 was taken from Prometheus example configurations, 2004 sound equally good.

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/output/PrometheusOutput.java:
```diff
@@ -0,0 +1,106 @@
+            SortedMap<String,Histogram> histograms, SortedMap<String,Meter> meters, SortedMap<String,Timer> timers )
+    {
+        // Prometheus does not support events, just output the latest event that occurred
+        String metricKey = gauges.firstKey();
```
this looks strange: we already registered export of registry in the init and here again during reporting we can potentially register some other set of events? how is that possible?

klaren replied:
This is for the cases where we do not register, `CHECK_POINT_DURATION` and `LOG_ROTATION_DURATION`. But I did not test how this affects logging if both csv and Prometheus is activated at the same time.

MishaDemianenko reviewed (commented):
couple of comments/questions, good otherwise

klaren committed:
Changed port of prometheus endpoint to 2004 as default.
Separated the registry for event-to-guage conversion to avoid pollution of the global registry.
Added more testing to prometheus endpoint to assure metrics can be registred after server starts.

klaren commented:
@MishaDemianenko New commit available for review.
